src/org/zaproxy/zap/extension/pscanrulesAlpha/XPoweredByHeaderInfoLeakScanner.java

```diff
@@ -58,30 +61,79 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
 	@Override
 	public void scanHttpResponseReceive(HttpMessage msg, int id, Source source) {
 		long start = System.currentTimeMillis();
-	
-		Vector<String> xpbOptions = msg.getResponseHeader().getHeaders(HEADER_NAME);
-		if (xpbOptions != null) { //Header Found
-			for (String xpbDirective : xpbOptions) {
-				Alert alert = new Alert(getPluginId(), Alert.RISK_LOW, Alert.CONFIDENCE_MEDIUM, //PluginID, Risk, Reliability
-					getName()); 
-		    		alert.setDetail(
-		    				getDescription(), //Description
-		    				msg.getRequestHeader().getURI().toString(), //URI
-		    				"",	// Param
-		    				"", // Attack
-		    				"", // Other info
-		    				getSolution(), //Solution
-		    				getReference(), //References
-		    				HEADER_NAME, // Evidence - Return the header name
-		    				200, // CWE Id
-		    				13,	// WASC Id
-		    				msg); //HttpMessage
-		    	parent.raiseAlert(id, alert);
-				}
+
+		List<String> xpbHeaders = new ArrayList<>();
+		if (isXPoweredByHeaderExist(msg)) {
+			xpbHeaders = getXPoweredByHeaders(msg);
+            raiseAlert(msg, id, xpbHeaders);
+			if (logger.isDebugEnabled()) {
+				logger.debug("\tScan of record " + id + " took " + (System.currentTimeMillis() - start) + " ms");
 			}
-		    if (logger.isDebugEnabled()) {
-		    	logger.debug("\tScan of record " + id + " took " + (System.currentTimeMillis() - start) + " ms");
-	    }
+        }
+	}
+
+	/**
+	 * Checks if there is any X-Powered-By header
+	 * @param msg Response Http message
+	 * @return boolean status of existence
+	 */
+	private boolean isXPoweredByHeaderExist(HttpMessage msg) {
+		return null != msg.getResponseHeader().getHeaders(HEADER_NAME);
+	}
+
+	/**
+	 * Extracts the list of "X-Powered-By" headers, and returns them without changing
+	 * their cases.
+	 * @param msg Response Http message
+	 * @return list of the matched headers
+	 */
+	private List<String> getXPoweredByHeaders(HttpMessage msg) {
+		List<String> matchedHeaders = new ArrayList<>();
+		String headers = msg.getResponseHeader().toString();
+		String[] headerElements = headers.split("\\r\\n");
+		Pattern pattern = Pattern.compile("X-Powered-By.*", Pattern.CASE_INSENSITIVE);
+		Matcher matcher = pattern.matcher(headers);
+		for (String header : headerElements) {
+			if (matcher.find()) {
+				String match = matcher.group();
+				matchedHeaders.add(match);
+			}
+		}
+		return matchedHeaders;
+	}
+
+    /**
+     * Raises an alert with the "Evidence" or "Other" field filled-in depending on the header repetition.
+     * @param msg The Http message containing the response headers
+     * @param id The alert ID
+     */
+	private void raiseAlert(HttpMessage msg, int id, List<String> xpbHeaders) {
+        String alertEvidence = xpbHeaders.get(0);
+        String alertOtherInfo = "";
+        if (xpbHeaders.size() > 1) { // we have multiple X-Powered-By headers
+            StringBuilder sb = new StringBuilder();

```

Comment:
It could set a higher capacity (the length of following string and the headers are more than the default).